File: qdisc/qdisc_parser.go
// qdisc parser a-la tc -s show qdisc

// Based on https://github.com/ema/qdisc, adapted for reusable objects, w/ data
// presented as slices. The latter allows for metrics generation in a loop.

package qdisc

import (
	"time"
)

const (
	// uint32 indices:
	QDISC_PARENT = iota
	QDISC_HANDLE
	QDISC_PACKETS
	QDISC_DROPS
	QDISC_REQUEUES
	QDISC_OVERLIMITS
	QDISC_QLEN
	QDISC_BACKLOG

	// Must be last:
	QDISK_UINT32_NUM_STATS
)

const (
	// uint64 indices:
	QDISC_BYTES = iota
	QDISC_GCFLOWS
	QDISC_THROTTLED
	QDISC_FLOWSPLIMIT

	// Must be last:
	QDISK_UINT64_NUM_STATS
)

const (
	// maj:min split:
	QDISC_MAJ_NUM_BITS = 16
	QDISC_MIN_NUM_BITS = 32 - QDISC_MAJ_NUM_BITS

	// How often to refresh the Interface index -> name cache:
	QDISC_IF_INDEX_TO_NAME_CACHE_REFRESH_INTERVAL = 60 * time.Second
)

type QdiscInfoKey struct {
	IfIndex uint32
	Handle  uint32
}

type QdiscInfo struct {
	IfName string
	Kind   string
	Uint32 [QDISK_UINT32_NUM_STATS]uint32
	Uint64 [QDISK_UINT64_NUM_STATS]uint64
	// Scan number used to identify out of scope interfaces:
	scanNum int
}

// QdiscStats object will be used for metrics generation as a current, previous
// tandem. The following information will be shared by both members; normally
// the access should be protected by a lock but since only one member is used at
// a time, the lock can be skipped.
type QdiscStatsShared struct {
	// Scan number used to identify out of scope qdiscs; incremented w/ every
	// call, qdiscs that have scan#(I/F) != scan#, will be removed:
	scanNum *int

	// Interface index -> name cache, refreshed periodically or every time there
	// is a miss:
	ifIndexToNameCache            map[uint32]string
	ifIndexToNameCacheLastRefresh time.Time

	// Netlink connection and request, however since the package that defines it
	// may be OS specific, use an unspecified type here:
	netConn       any
	netReqMessage any
}

type QdiscStats struct {
	// Map info by (ifIndex, handle), since it is unique:
	Info map[QdiscInfoKey]*QdiscInfo

	// Shared info:
	shared *QdiscStatsShared
}

func NewQdiscStats() *QdiscStats {
	return &QdiscStats{
		Info: make(map[QdiscInfoKey]*QdiscInfo),
		shared: &QdiscStatsShared{
			ifIndexToNameCache: make(map[uint32]string),
			scanNum:            new(int),
		},
	}
}

func (qs *QdiscStats) Clone() *QdiscStats {
	newQs := &QdiscStats{
		Info:   make(map[QdiscInfoKey]*QdiscInfo),
		shared: qs.shared,
	}

	for qiKei, qi := range qs.Info {
		newQs.Info[qiKei] = &QdiscInfo{
			IfName:  qi.IfName,
			Kind:    qi.Kind,
			scanNum: qi.scanNum,
		}
	}

	return newQs
}

File: tools/test/qdisc/qdisc_parser.py
#! /usr/bin/env python3

from dataclasses import dataclass, field
from typing import List, Optional

# JSON serialize-able QdiscStats, matching qdisc/qdisc_parser.go:

# uint32 indexes:
QDISC_PARENT = 0
QDISC_HANDLE = 1
QDISC_PACKETS = 2
QDISC_DROPS = 3
QDISC_REQUEUES = 4
QDISC_OVERLIMITS = 5
QDISC_QLEN = 6
QDISC_BACKLOG = 7
QDISK_UINT32_NUM_STATS = 8

# uint64 indexes:
QDISC_BYTES = 0
QDISC_GCFLOWS = 1
QDISC_THROTTLED = 2
QDISC_FLOWSPLIMIT = 3
QDISK_UINT64_NUM_STATS = 4

QDISC_MAJ_NUM_BITS = 16
QDISC_MIN_NUM_BITS = 32 - QDISC_MAJ_NUM_BITS


def format_maj_min(val: int) -> str:
    return ":".join(
        [
            f"{val >> QDISC_MIN_NUM_BITS:0{(QDISC_MAJ_NUM_BITS + 3)//4}x}",
            f"{val & ((1 << QDISC_MIN_NUM_BITS) - 1):0{(QDISC_MIN_NUM_BITS + 3)//4}x}",
        ]
    )


@dataclass
class QdiscInfoKey:
    IfIndex: int = 0
    Handle: int = 0

    def __str__(self) -> str:
        return f"QdiscInfoKey(IfIndex={self.IfIndex}, Handle={self.Handle} ({format_maj_min(self.Handle)}))"

    def __hash__(self) -> int:
        return (self.IfIndex << 32) + self.Handle


@dataclass
class QdiscInfo:
    IfName: Optional[str] = None
    Kind: Optional[str] = None
    Uint32: List[int] = field(default_factory=lambda: [0] * QDISK_UINT32_NUM_STATS)
    Uint64: List[int] = field(default_factory=lambda: [0] * QDISK_UINT64_NUM_STATS)
